assets/js/blocks/cart-checkout/payment-methods/saved-payment-method-options.js

```diff
@@ -25,139 +22,91 @@ import { getPaymentMethods } from '@woocommerce/blocks-registry';
-			method.gateway
-		),
-		name: `wc-saved-payment-method-token-${ tokenId }`,
-		onChange: ( token ) => {
-			const savedTokenKey = `wc-${ method.gateway }-payment-token`;
-			setActivePaymentMethod( method.gateway );
-			setPaymentStatus().started( {
-				payment_method: method.gateway,
-				[ savedTokenKey ]: token + '',
-				isSavedToken: true,
-			} );
-		},
-	};
+const getDefaultLabel = ( { method } ) => {
+	return sprintf(
+		/* translators: %s is the name of the payment method gateway. */
+		__( 'Saved token for %s', 'woo-gutenberg-products-block' ),
+		method.gateway
+	);
 };
 
 const SavedPaymentMethodOptions = () => {
 	const {
-		setPaymentStatus,
 		customerPaymentMethods,
 		activePaymentMethod,
 		setActivePaymentMethod,
 		activeSavedToken,
-		setActiveSavedToken,
 	} = usePaymentMethodDataContext();
-	const standardMethods = getPaymentMethods();
 	const { paymentMethods } = usePaymentMethods();
 	const paymentMethodInterface = usePaymentMethodInterface();
+	const { noticeContexts } = useEmitResponse();
+	const { removeNotice } = useStoreNotices();
+	const { dispatchCheckoutEvent } = useStoreEvents();
 
-	/**
-	 * @type      {Object} Options
-	 * @property  {Array}  current  The current options on the type.
-	 */
-	const currentOptions = useRef( [] );
-
-	const updateToken = useCallback(
-		( token ) => {
-			setActiveSavedToken( token );
-		},
-		[ setActiveSavedToken ]
-	);
-
-	useEffect( () => {
+	const options = useMemo( () => {
 		const types = Object.keys( customerPaymentMethods );
-		const options = types
+		return types
 			.flatMap( ( type ) => {
 				const typeMethods = customerPaymentMethods[ type ];
 				return typeMethods.map( ( paymentMethod ) => {
-					const option =
-						type === 'cc' || type === 'echeck'
-							? getCcOrEcheckPaymentMethodOption(
-									paymentMethod,
-									setActivePaymentMethod,
-									setPaymentStatus
-							  )
-							: getDefaultPaymentMethodOptions(
-									paymentMethod,
-									setActivePaymentMethod,
-									setPaymentStatus
-							  );
-					if (
-						! activePaymentMethod &&
-						paymentMethod.is_default &&
-						activeSavedToken === ''
-					) {
-						updateToken( paymentMethod.tokenId + '' );
-						option.onChange( paymentMethod.tokenId );
-					}
-					return option;
+					const isCC = type === 'cc' || type === 'echeck';
+					const paymentMethodSlug = paymentMethod.method.gateway;
+					return {
+						name: `wc-saved-payment-method-token-${ paymentMethodSlug }`,
+						label: isCC
+							? getCcOrEcheckLabel( paymentMethod )
+							: getDefaultLabel( paymentMethod ),
+						value: paymentMethod.tokenId.toString(),
+						onChange: ( token ) => {
+							const savedTokenKey = `wc-${ paymentMethodSlug }-payment-token`;
+							setActivePaymentMethod( paymentMethodSlug, {
+								token,
+								payment_method: paymentMethodSlug,
+								[ savedTokenKey ]: token.toString(),
+								isSavedToken: true,
+							} );
+							removeNotice(
```

Comment:
This was previously handled in the context provider via useEffect, but because we want the token and method ID it seemed best to move it all to the onClick handler.